Patch and ArrayMember classes

pblivingston opened the pull request:
Closes #8 

## What

Adds ArrayMember classes to provide a clean interface for handling array-like properties with:
- .Set($val)
- .Get()

Adds Patch class with:
- Patch.asio[i]
- Patch.OutA2[i]-OutA5[i]
- Patch.composite[i]
- Patch.insert[i]
- Patch.postFaderComposite
- Patch.postFxInsert

## Notes

Tests pass for all kinds.

.Set($val) and .Get() on
- asio[i] (all kinds)
- OutA2[i]-OutA5[i] (Banana, Potato)
require an asio device so were tested manually.

pblivingston committed:
ArrayMember classes

pblivingston committed:
Patch

Patch class with:
- Patch.asio[i]
- Patch.OutA2[i]-OutA5[i]
- Patch.composite[i]
- Patch.insert[i]
- Patch.postFaderComposite
- Patch.postFxInsert

pblivingston committed:
Update kinds.ps1

pblivingston committed:
tests

pester tests pass

pblivingston committed:
Merge branch 'iremote' into patch-arraymembers

pblivingston committed:
update docs

manual tests all pass:
- asio[i].set($val)
- asio[i].get()
- outa2[i]-outa5[i].set($val)
- outa2[i]-outa5[i].get()

these require an asio device

onyx-and-iris commented on README.md:
Did you mean to include {Patch} in Patch.postFxInsert?

The examples are missing the patch class. i.e. $vmr.asio[3].set(2)  -> $vmr.patch.asio[3].set(2)

pblivingston replied:
Corrected this in [f3ed1de](https://github.com/onyx-and-iris/voicemeeter-api-powershell/pull/9/commits/f3ed1de55781b9732440ed06eac8eccf8c40bbe9) and made a slight change for readability.

onyx-and-iris commented:
I've noticed on reviewing this PR that it should be possible to control OUTA2 ASIO patching for basic kind which means we can update the basic KindMap:

```powershell
    'basic'  = @{
        ...
        'asio_in'   = 4
        'asio_out'  = 8
```

I'm not entirely clear on this but the basic GUI marks A as `main out` and B as `virtual out` but A1 and A2 can be assigned to physical devices which matters for this ASIO patching. So we could update `AddASIOOutMembers` like so:

```powershell
    hidden [void] AddASIOOutMembers () {
        $num_A = $this.remote.kind.p_out
        if ($this.remote.kind.name -eq 'basic') {
            $num_A += $this.remote.kind.v_out
        }
        $asio_out = $this.remote.kind.asio_out

        ...
```

This allows us to control OUTA2 ASIO patching for basic kind using this module.

Let me know your thoughts.

pblivingston committed:
Update README.md

- correct postfxinsert
- correct examples
- better readability for patch arraymembers

pblivingston committed:
add a2 for basic

patch.outa2[i]

pblivingston commented:
Yeah, that looks good! Latest commit adds the change. Manual tests pass on my end.

I just did a quick test of bus[1].device.wdm, etc. through macrobuttons, which also looks like it works. I was already planning to decouple Device from Bus and Strip, so I could do that next and add Device to the virtual bus for Basic.

Let me know if you'd like me to hold off on that for any reason.

onyx-and-iris commented:
Feel free to decouple Device class next.
